From 104c20821e4cfdfb46367154753e58e65984856c Mon Sep 17 00:00:00 2001
From: James <aslan.jnn@gmail.com>
Date: Mon, 3 Oct 2022 16:01:27 +1100
Subject: Activate force selection when mouse reporting is disabled

---
 src/browser/services/SelectionService.ts | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/browser/services/SelectionService.ts b/src/browser/services/SelectionService.ts
index 486c1941..3970372b 100644
--- a/src/browser/services/SelectionService.ts
+++ b/src/browser/services/SelectionService.ts
@@ -427,6 +427,10 @@ export class SelectionService extends Disposable implements ISelectionService {
    * @param event The mouse event.
    */
   public shouldForceSelection(event: MouseEvent): boolean {
+    if (!this._optionsService.rawOptions.allowMouseReporting) {
+      return true;
+    }
+
     if (Browser.isMac) {
       return event.altKey && this._optionsService.rawOptions.macOptionClickForcesSelection;
     }
--
2.37.1 (Apple Git-137.1)
